common/models/checkpoint.js

```diff
@@ -27,43 +28,59 @@ module.exports = function(Checkpoint) {
    * Get the current checkpoint id
    * @callback {Function} callback
    * @param {Error} err
-   * @param {Number} checkpointId The current checkpoint id
+   * @param {Number} checkpoint The current checkpoint seq
    */
-
   Checkpoint.current = function(cb) {
     var Checkpoint = this;
-    this.find({
-      limit: 1,
-      order: 'seq DESC'
-    }, function(err, checkpoints) {
-      if (err) return cb(err);
-      var checkpoint = checkpoints[0];
-      if (checkpoint) {
-        cb(null, checkpoint.seq);
-      } else {
-        Checkpoint.create({ seq: 1 }, function(err, checkpoint) {
-          if (err) return cb(err);
-          cb(null, checkpoint.seq);
-        });
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (cp) {
+          cb(null, cp.seq);
+        } else {
+          Checkpoint.init(function(err, data) {
```

Comment:
I am little bit afraid that this can create a race condition - what if there are two calls of `Checkpoint.current` called at the same time? Both will run `findOne` and get back `null`, and then they will go ahead and create two new checkpoints.

This test may be able to reproduce the race condition (or not, it depends on memory connector internal details). If it does, then it should be added to the test suite:

```js
async.parallel([
  function(next) { Checkpoint.current(next); },
  function(next) { Checkpoint.current(next); }
], function(err, list) {
  if (err) return done(err);
  Checkpoint.find(function(err) {
    if (err) return done(err);
    expect(list).to.have.length(1);
    done();
  });
});
```


I am proposing to use `findOrCreate` instead. Here is a mockup implementation:

```js
Checkpoint.current = function(cb) {
  var Checkpoint = this;
  var query = { limit: 1 }; // match all instances, return only one
  var initialData = { seq: 1 };
  Checkpoint.findOrCreate(query, initialData, function(err, list) {
    if (err) return cb(err);
    if (list.length != 1) {
     err = new Error('Assertion failed, there are ' + list.length + ' Checkpoint instances. Expected exactly 1.');
     return cb(err);
    }
    cb(list[0].seq);
  });
};
```

Thoughts?

The [memory connector](https://github.com/strongloop/loopback-datasource-juggler/blob/dfda177ff7a2f3f6dce924e00c479b546f68f420/lib/connectors/memory.js) does not implement atomic `findOrCreate` now, we will need to fix that as part of this work too. As part of that, this [test setup](https://github.com/strongloop/loopback-datasource-juggler/blob/dfda177ff7a2f3f6dce924e00c479b546f68f420/test/memory.test.js#L815-L823) should be fixed to not provide a dummy implementation (that is actually not atomic!).